File: vripper-server/src/main/java/me/vripper/services/DownloadSpeedService.java
package me.vripper.services;

import java.util.concurrent.atomic.AtomicLong;
import lombok.Getter;
import me.vripper.event.Event;
import me.vripper.event.EventBus;
import org.springframework.scheduling.annotation.EnableScheduling;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

@Service
@EnableScheduling
public class DownloadSpeedService {

  private final AtomicLong read = new AtomicLong(0);
  private final EventBus eventBus;
  @Getter private long currentValue;
  private boolean allowWrite = false;

  public DownloadSpeedService(EventBus eventBus) {
    this.eventBus = eventBus;
  }

  public void increase(long read) {
    if (allowWrite) {
      this.read.addAndGet(read);
    }
  }

  @Scheduled(fixedDelay = 1000)
  private void calc() {
    allowWrite = false;
    long newValue = read.getAndSet(0);
    if (newValue != currentValue) {
      currentValue = newValue;
      eventBus.publishEvent(Event.wrap(Event.Kind.BYTES_PER_SECOND, currentValue));
    }
    allowWrite = true;
  }
}

File: vripper-core/src/main/kotlin/me/vripper/services/DownloadSpeedService.kt
package me.vripper.services

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.filterIsInstance
import me.vripper.event.DownloadSpeedEvent
import me.vripper.event.EventBus
import me.vripper.event.QueueStateEvent
import me.vripper.model.DownloadSpeed
import me.vripper.utilities.LoggerDelegate
import java.util.concurrent.atomic.AtomicLong

internal class DownloadSpeedService(
    private val eventBus: EventBus,
) {

    companion object {
        const val DOWNLOAD_POLL_RATE = 2500
    }

    private val log by LoggerDelegate()
    private val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val bytesCount = AtomicLong(0)
    private var job: Job? = null
    private var queueStateUpdateJob: Job? = null

    fun init() {
        queueStateUpdateJob?.cancel()
        queueStateUpdateJob = coroutineScope.launch {
            eventBus.events.filterIsInstance(QueueStateEvent::class).collect {
                if (it.queueState.running + it.queueState.remaining > 0) {
                    if (job == null || job?.isActive == false) {
                        job = coroutineScope.launch {
                            eventBus.publishEvent(DownloadSpeedEvent(DownloadSpeed(0L)))
                            while (isActive) {
                                delay(DOWNLOAD_POLL_RATE.toLong())
                                val newValue = bytesCount.getAndSet(0)
                                val speed = DownloadSpeed(((newValue * 1000) / DOWNLOAD_POLL_RATE))
                                log.debug(
                                    "[{}] Publishing event: DownloadSpeedEvent({})",
                                    System.currentTimeMillis(),
                                    speed
                                )
                                eventBus.publishEvent(DownloadSpeedEvent(speed))
                            }
                        }
                    }
                } else {
                    job?.cancel()
                    coroutineScope.launch {
                        delay(DOWNLOAD_POLL_RATE + 500L)
                        log.debug("[{}] Publishing event: DownloadSpeedEvent(0)", System.currentTimeMillis())
                        eventBus.publishEvent(DownloadSpeedEvent(DownloadSpeed(0L)))
                    }
                }
            }
        }
    }

    fun halt() {
        queueStateUpdateJob?.cancel()
        job?.cancel()
    }

    fun reportDownloadedBytes(count: Long) {
        bytesCount.addAndGet(count)
    }
}
